[contacts][next] Add Contacts@Next

Wenszel opened the pull request:
# Why

The old module was written using `Map/Bundle`, but now the same logic can be implemented much better with `Record`. This change also enabled us to easily extend the API with granular modification functions (e.g. `contact.addPhone()`) and create a very flexible `.patch()` method.

What has changed:

* Migration from `Map/Bundle` to `Record`
* New object-oriented API, focused on easier contact modification (e.g., `addPhone`, `getPhones`, `deletePhone`, `updatePhone`, `patch`, `setFirstName`)
* Significant performance optimization for fetching contacts with pagination.
* Added the `PartialContactDetails` data type, making `getDetails` type-safe based on the requested fields. 
* Refactored code by introducing an abstraction over `ContactsContract` and clearly separating DTO from Domain
* Migration from promises to coroutines to handle intents
* Refactored permission handling

# How

### New API

Creating a contact:

```js
const contact = await Contact.create({ givenName: "Andrew", lastName: "Jones" })
```

Granular functions for managing properties:

```js
const emailId = await contact.addEmail({
  address: "andrew@foo.bar",
  label: "work",
})

await contact.updateEmail({ id: emailId, address: "luke@foo.bar", label: "home" })
await contact.deleteEmail(emailId)
```

Patch:

* distinguishes between existing and new properties
* updates only specified fields
* unlike the old module, it doesn’t wipe all data just to update the contact

```js
await contact.patch({
  givenName: "Tom",
  lastName: "Smith",
  company: null,
  emails: [
    {
      id: emailId,
      address: "tom.smith@foo.bar",
      label: "work",
    },
    {
      address: "new.tom@foo.bar",
      label: "home",
    },
  ],
})
```

Fetching details:

```js
await contact.getDetails([ContactField.EMAILS, ContactField.GIVEN_NAME])
await Contact.getAllWithDetails([ContactField.EMAILS, ContactField.GIVEN_NAME]) 
```

Moreover, thanks to the `PartialContactDetails` data type, TypeScript knows exactly what is inside the details object:

<img width="398" height="152" alt="Screenshot 2025-12-02 at 12 58 14" src="https://github.com/user-attachments/assets/6b33b483-68c0-4ad1-a4e1-6732440c7a55" />

### Performance improvement 
Due to the monolithic nature of the `Data` table in Android Contacts, creating efficient queries with limit/offset is difficult. Because of this, I utilized a two-step query approach, which resulted in a significant performance improvement. In the old module, the query didn't scale with the database size when fetching contacts with a limit; now it scales perfectly.

| Rows in DB | Old (ms) | New (ms) | Diff (x) |
| :--- | :--- | :--- | :--- |
| 100 | 23 | 21.1 | 1.09x |
| 1000 | 179.2 | 41.4 | 4.33x |
| 5000 | 670.6 | 46.7 | 14.36x |
| 10000 | 1579.3 | 40.8 | 38.71x |

### Migration to Record (from Map/Bundle)

Thanks to new features introduced to `Record`, it became feasible to migrate completely from `Map/Bundle`. `Map/Bundle` is extremely flexible in terms of what can be passed to native code and returned to JS, but this flexibility comes with a downside — the structure of objects is not standardized, which forces constant casting.

The most important classes that enabled the migration:

* **`RecordFormatter`** — lets us decide which fields should be sent to the JS side. No more objects filled with nulls.
* **`ValueOrUndefined`** — distinguishes between explicitly passing `null` and leaving a field `undefined`. This matters when patching — we need to know whether the user wants to change a property or keep it unchanged.
* **`Either`** — previously, `Map` let us pass any type. `Either` brings the same flexibility but keeps it within a typed environment.

### Logic Refactor

The API object structure differs from the `ContactContract` representation.  
That’s why I split the code into two domains: **Records (DTO layer)** and **Domain (abstraction over ContactsContract)**.

The abstraction looks like this:

* Each data type stored in the Contacts `Data` table (identified by a MIME type) is represented by an abstract `model` class.
* Model classes are extended by operation-specific types (`Existing`, `Appendable`, `New`, `Patch`). Each implements interfaces defining how database operations should be executed.
* Data extraction uses the `ExtractableField` interface, which defines how to build the projection and how to read values from a cursor.
* At the core is the `ContactRepository`, which takes objects implementing these interfaces and performs operations via `contentResolver`.

### Migration from Intents to Coroutines

Thanks to this change, we no longer create pending promises in one function and resolve them in another.  
Instead, we use `AppContextActivityResultLauncher`, register them in `RegisterActivityContracts`, and launch them wherever we want.


### Permissions Refactor

Similar to `MediaLibrary@Next`, I changed how permission exceptions are handled.

The only moment where explicit permission checking matters is when creating a `Contact` object — there’s no need to check permissions on every method call afterward.

Because of that, `ensurePermissions` was replaced by a `contentResolver.query` wrapper that catches security exceptions.

## Test plan

Adds tests to bare-expo

Wenszel committed:
Add Contacts@Next

lukmccall commented on apps/test-suite/tests/ContactsNext.ts:
```diff
@@ -0,0 +1,911 @@
+
+    t.it('should set and get familyName', async () => {
+      const newFamilyName = 'Doe';
+      console.log('contactId', contact);
```
```suggestion
```

lukmccall commented on apps/test-suite/tests/ContactsNext.ts:
```diff
@@ -0,0 +1,911 @@
+      t.expect(retrieved).toBeNull();
+    });
+  });
+  t.describe('Add, Get, Update, Delete list properties', () => {
```
```suggestion

  t.describe('Add, Get, Update, Delete list properties', () => {
```

lukmccall commented on packages/expo-contacts/android/src/main/java/expo/modules/contacts/next/domain/model/contact/NewContact.kt:
```diff
@@ -0,0 +1,19 @@
+
+data class NewContact(val modelsToInsert: List<Insertable>) {
+  fun toInsertOperations(): ArrayList<ContentProviderOperation> {
+    val output = arrayListOf<ContentProviderOperation>()
```
```suggestion
  fun toInsertOperations(): List<ContentProviderOperation> {
    val output = mutableListOf<ContentProviderOperation>()
```
If you have to use `ArrayList` here, I'm okay with it. However, always prefer to use `listOf`. Also, you don't need to use `mutableListOf` in this case. I would say that the `buildList` function is a better choice here - https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/build-list.html

lukmccall commented on packages/expo-contacts/android/src/main/java/expo/modules/contacts/next/domain/model/nickname/NicknameLabel.kt:
```diff
@@ -0,0 +1,42 @@
+
+sealed class NicknameLabel {
+  abstract val type: Int
+  abstract val label: String?
```
```suggestion
  open val label: String? = null
```

lukmccall commented on packages/expo-contacts/android/src/main/java/expo/modules/contacts/next/domain/model/website/WebsiteLabel.kt:
```diff
@@ -0,0 +1,50 @@
+
+sealed class WebsiteLabel {
+  abstract val type: Int
+  abstract val label: String?
```
```suggestion
  abstract val label: String
```

lukmccall commented on packages/expo-contacts/android/src/main/java/expo/modules/contacts/next/domain/query/QueryAggregator.kt:
```diff
@@ -0,0 +1,100 @@
+  companion object {
+    fun aggregate(cursor: Cursor, extractors: Set<ExtractableField<*>>): List<ExistingContact> {
+      val extractorsByMimeType = extractors.associateBy { it.mimeType }
+      val contactsMap = mutableMapOf<String, ContactModelBuilder>()
```
Consider using `buildMap` - https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/build-map.html

lukmccall commented on packages/expo-contacts/android/src/main/java/expo/modules/contacts/next/domain/query/QueryAggregator.kt:
```diff
@@ -0,0 +1,100 @@
+    }
+
+    fun <T : Extractable> aggregateOneField(cursor: Cursor, extractor: ExtractableField<T>): List<T> {
+      val foundItems = mutableListOf<T>()
```
Consider using `buildList`

Wenszel committed:
Apply PR comments

Wenszel committed:
Make compatible with iOS

Wenszel committed:
Add isFavourite, displayName

Wenszel committed:
Add pagination and make all tests pass

lukmccall commented on apps/test-suite/tests/ContactsNext.ts:
```diff
@@ -0,0 +1,1561 @@
+      t.expect(contactsInGroup.some((c) => c.id === contact1.id)).toBe(true);
+      t.expect(contactsInGroup.some((c) => c.id === contact2.id)).toBe(true);
+    });
+    t.it('.getContacts({name: string} should correctly filter contacts by name)', async () => {
```
```suggestion

    t.it('.getContacts({name: string} should correctly filter contacts by name)', async () => {
```

lukmccall commented on packages/expo-contacts/android/src/main/java/expo/modules/contacts/next/domain/model/photo/PhotoModel.kt:
```diff
@@ -0,0 +1,15 @@
+    ContentValues().apply {
+      put(Photo.PHOTO, photo)
+    }
+}
```
```suggestion
}

```

lukmccall commented on packages/expo-contacts/android/src/main/java/expo/modules/contacts/next/services/properties/SingleDataProperty.kt:
```diff
@@ -0,0 +1,43 @@
+      .firstOrNull()
+      .ifNull {
+        return null
+      }
```
```suggestion
      .firstOrNull()
      ?: return null
```

lukmccall commented on packages/expo-contacts/android/src/main/java/expo/modules/contacts/next/services/ImageByteArrayConverter.kt:
```diff
@@ -0,0 +1,18 @@
+    bitmap.compress(Bitmap.CompressFormat.JPEG, 80, stream)
+    return stream.toByteArray()
+  }
+}
```
```suggestion
}

```

lukmccall reviewed (commented):
Looks good!

Wenszel committed:
Refactor properties

Wenszel committed:
Add hasAny and getCount

Wenszel committed:
Lint

Wenszel committed:
Refactor TypeScript

Wenszel committed:
Fix TestModules

Wenszel committed:
PR Comments

Wenszel committed:
Add changelog

Wenszel committed:
Add new line in tests

Wenszel committed:
Lint tests

Wenszel committed:
Yarn build

Wenszel committed:
Lint import order

Wenszel committed:
Android spotless lint

Wenszel committed:
Fix labels

Wenszel committed:
Remove iOS module from config for CI

Wenszel committed:
Rename fallback label to other

lukmccall commented on packages/expo-contacts/android/src/main/java/expo/modules/contacts/next/domain/model/email/EmailLabel.kt:
```diff
@@ -0,0 +1,28 @@
+import android.provider.ContactsContract.CommonDataKinds.Email
+
+sealed class EmailLabel {
+  abstract val type: Int
```
Move to constructor

Wenszel committed:
Fix getDetails

Wenszel committed:
Refactor labels

lukmccall commented on packages/expo-contacts/android/src/main/java/expo/modules/contacts/next/domain/model/website/WebsiteLabel.kt:
```diff
@@ -0,0 +1,40 @@
+import android.provider.ContactsContract.CommonDataKinds.Website
+
+sealed class WebsiteLabel {
+  abstract val type: Int
```
Can be moved to the constructor

Wenszel committed:
Change patch behaviour

Wenszel committed:
Fix presentCreateForm and remove ClearableField

Wenszel committed:
Remove unused fields

Wenszel committed:
Lint

Wenszel committed:
Fix test getDetails

Wenszel committed:
Change date to number

Wenszel committed:
Fix check packages

Wenszel committed:
Disable filtering by contactIds when limit and offset are null

Wenszel committed:
[iOS][contacts] Add Contacts@Next (#41278)

# Why
depends on: [#40886](https://github.com/expo/expo/pull/40886)
This PR implements features for iOS that were previously added for
Android.
Additionally:
- Adds shared objects: `Group`, `Container` 
- Adds support for `NonGregorianBirthday` and `SocialProfiles`,
`Nickname`, `MaidenName`
- Fixes fetching contacts from a group/container with filtering by name,
which did not work correctly in the previous version.

# Test Plan
Bare-expo test suite

Wenszel committed:
Fix starred

Wenszel committed:
Fix cloud accounts

Wenszel committed:
[docs][contacts] Add docs for Contacts@Next (#41557)

# Why
Adds docs for Contacts@Next 

# How
Adds `contacts-next.mdx` and docs in typescript files.

# Test Plan
Tested on the docs app

Wenszel committed:
Rename phoneticOrganizationName
